File: ng/macos/Tests/JiaoWoXuanCoreTests/CoreTests.swift
import Foundation
import XCTest
@testable import JiaoWoXuanCore

final class ModelDecodingTests: XCTestCase {
    func testGroupConflictsKeepRawDictionaryKeys() throws {
        let json = """
        {"name":"物理","is_pe":false,"priority":["A1_b2","held"],"held":"held","held_label":"大学物理 - 02",
         "watched_count":1,"fatal":false,"members":[],
         "conflicts":{"A1_b2":{"status":"conflict","with":"高数 - 01","detail":"周一 第2节 (第1周)"}},
         "conflict_count":1}
        """
        let group = try JSONDecoder().decode(PriorityGroup.self, from: Data(json.utf8))
        XCTAssertEqual(group.conflicts["A1_b2"]?.status, .conflict)
        XCTAssertEqual(group.conflicts["A1_b2"]?.note, "与已选 高数 - 01 冲突：周一 第2节 (第1周)")
        XCTAssertEqual(group.plan, PriorityGroup.Plan(name: "物理", isPe: false, priority: ["A1_b2", "held"]))
    }

    func testRatingZeroScoreIsKept() throws {
        let json = #"{"status":"rated","score":0,"count":0,"teacher":null,"semester":null,"updated_at":null,"message":null}"#
        let rating = try JSONDecoder().decode(CourseRating.self, from: Data(json.utf8))
        XCTAssertEqual(rating.score, 0)
        XCTAssertEqual(rating.count, 0)
        XCTAssertEqual(rating.scoreText, "0.0")
    }
}

final class CourseFilterTests: XCTestCase {
    private func course(_ id: String, _ title: String, score: Double? = nil, open: Bool = false, group: String? = nil) -> CourseRow {
        CourseRow(
            jxbId: id, title: title, className: "01", summary: "", detail: "", teachers: "", schedule: [],
            locations: [], searchText: "\(title) \(id)".lowercased(), seatText: "-",
            availability: open ? .open : .full, availabilityText: "", group: group, chosen: false, category: "主修",
            ratingText: "", rating: CourseRating(status: score == nil ? .unknown : .rated, score: score),
            kch: nil, sksj: nil
        )
    }

    func testRatingSortPutsUnratedLast() {
        var filter = CourseFilter()
        filter.sort = .rating
        let result = filter.apply(to: [course("a", "甲"), course("b", "乙", score: 3), course("c", "丙", score: 4.5)])
        XCTAssertEqual(result.map(\.jxbId), ["c", "b", "a"])
    }

    func testFiltersCombine() {
        var filter = CourseFilter()
        filter.onlyOpen = true
        filter.onlyUnassigned = true
        filter.query = "物理"
        let result = filter.apply(to: [
            course("a", "大学物理", open: true),
            course("b", "大学物理", open: true, group: "g"),
            course("c", "大学物理"),
            course("d", "高等数学", open: true),
        ])
        XCTAssertEqual(result.map(\.jxbId), ["a"])
    }
}

extension CourseRating {
    init(status: RatingStatus, score: Double?) {
        self.init(status: status, score: score, count: nil, teacher: nil, semester: nil, updatedAt: nil, message: nil)
    }
}

final class BackendClientTests: XCTestCase {
    /// 真正起一个 ng_service(需要仓库里的 conda Python),走一次 hello/snapshot/错误路径。
    func testRoundTripAgainstRealService() async throws {
        let repo = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent()
            .deletingLastPathComponent().deletingLastPathComponent()
        guard FileManager.default.fileExists(atPath: repo.appendingPathComponent("ng_service.py").path) else {
            throw XCTSkip("仓库外运行")
        }
        let data = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        try FileManager.default.createDirectory(at: data, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: data) }

        var env = ProcessInfo.processInfo.environment
        env["SJTU_MONITOR_ROOT"] = repo.path
        env["SJTU_MONITOR_DATA_DIR"] = data.path
        let launch: BackendLaunch
        do {
            launch = try BackendLocator.resolve(environment: env)
        } catch {
            throw XCTSkip("没有可用的 Python: \(error)")
        }
        var launchEnv = launch.environment
        launchEnv["SJTU_MONITOR_DATA_DIR"] = data.path
        let configured = BackendLaunch(executable: launch.executable, arguments: launch.arguments,
                                       environment: launchEnv, workingDirectory: launch.workingDirectory,
                                       description: launch.description)

        let ready = expectation(description: "ready")
        let client = BackendClient(launch: configured) { event in
            if case .ready = event { ready.fulfill() }
        }
        try client.start()
        await fulfillment(of: [ready], timeout: 30)

        let hello = try await client.call("hello", as: HelloInfo.self)
        XCTAssertEqual(hello.protocolVersion, 1)
        XCTAssertEqual(URL(fileURLWithPath: hello.dataDir).standardizedFileURL.path,
                       data.standardizedFileURL.path)

        let snapshot = try await client.call("snapshot", as: Snapshot.self)
        XCTAssertFalse(snapshot.onboarding.completed)
        XCTAssertTrue(snapshot.courses.isEmpty)

        do {
            _ = try await client.call("no.such.method", as: OkResult.self)
            XCTFail("应当报错")
        } catch let error as BackendError {
            XCTAssertEqual(error.code, "method_not_found")
        }
        client.shutdown()
        XCTAssertFalse(client.isRunning)
    }
}

final class RealSnapshotTests: XCTestCase {
    /// NG_SNAPSHOT_JSON 指向一份 ng_service snapshot 的输出时,校验真实数据能完整解码。
    func testDecodesRealSnapshotWhenProvided() throws {
        guard let path = ProcessInfo.processInfo.environment["NG_SNAPSHOT_JSON"] else {
            throw XCTSkip("未提供 NG_SNAPSHOT_JSON")
        }
        let snapshot = try JSONDecoder().decode(Snapshot.self, from: Data(contentsOf: URL(fileURLWithPath: path)))
        XCTAssertFalse(snapshot.courses.isEmpty)
        var filter = CourseFilter()
        filter.sort = .rating
        XCTAssertEqual(filter.apply(to: snapshot.courses).count, snapshot.courses.count)
    }
}

File: ng/windows/JiaoWoXuan.Core.Tests/CoreTests.cs
using System.Collections.Concurrent;
using System.Text.Json;
using JiaoWoXuan.Core;
using Xunit;

namespace JiaoWoXuan.Core.Tests;

public class ModelDecodingTests
{
    [Fact]
    public void GroupConflictsKeepRawDictionaryKeys()
    {
        const string json = """
            {"name":"物理","is_pe":false,"priority":["A1_b2","held"],"held":"held","held_label":"大学物理 - 02",
             "watched_count":1,"fatal":false,"members":[],
             "conflicts":{"A1_b2":{"status":"conflict","with":"高数 - 01","detail":"周一 第2节 (第1周)"}},
             "conflict_count":1}
            """;
        var group = JsonSerializer.Deserialize<PriorityGroup>(json, Json.Options)!;
        Assert.True(group.Conflicts["A1_b2"].IsConflict);
        Assert.Equal("与已选 高数 - 01 冲突：周一 第2节 (第1周)", group.Conflicts["A1_b2"].Note);
        Assert.Equal(group.PlanSignature, group.Clone().PlanSignature);
    }

    [Fact]
    public void SnakeCaseEnumsDecode()
    {
        var metrics = JsonSerializer.Deserialize<Metrics>(
            """{"queries":1,"groups":1,"snapshot":1,"watched":1,"open_courses":2,"interval":"60-120s","auto_swap":"dry_run"}""",
            Json.Options)!;
        Assert.Equal(AutoSwapState.DryRun, metrics.AutoSwap);
        var rating = JsonSerializer.Deserialize<CourseRating>(
            """{"status":"teacher_unrated","score":0,"count":0}""", Json.Options)!;
        Assert.Equal(RatingStatus.TeacherUnrated, rating.Status);
        Assert.Equal(0, rating.Score);
        Assert.Null(rating.SortScore);
    }

    [Fact]
    public void BootstrapResultSiteTermLabel()
    {
        var exit = JsonSerializer.Deserialize<ProcessExit>(
            """{"task":"detect-term","code":0,"stopped":false,"result":{"ok":true,"site_term":{"label":"2026-2027 秋"}}}""",
            Json.Options)!;
        Assert.Equal("2026-2027 秋", exit.Result!.SiteTermLabel);
    }
}

public class CourseFilterTests
{
    static CourseRow Course(string id, string title, double? score = null, bool open = false, string? group = null) => new()
    {
        JxbId = id,
        Title = title,
        SearchText = $"{title} {id}".ToLowerInvariant(),
        Availability = open ? Availability.Open : Availability.Full,
        Group = group,
        Rating = new CourseRating { Status = score is null ? RatingStatus.Unknown : RatingStatus.Rated, Score = score },
    };

    [Fact]
    public void RatingSortPutsUnratedLast()
    {
        var result = new CourseFilter()
            .Apply([Course("a", "甲"), Course("b", "乙", 3), Course("c", "丙", 4.5)], new CourseSortOrder(CourseColumn.Rating, Descending: true));
        Assert.Equal(["c", "b", "a"], result.Select(c => c.JxbId));
    }

    [Fact]
    public void HeaderClickCyclesAscendingDescendingOff()
    {
        var sort = new CourseSortOrder().Toggle(CourseColumn.Title);
        Assert.Equal(new CourseSortOrder(CourseColumn.Title), sort);
        sort = sort.Toggle(CourseColumn.Title);
        Assert.True(sort.Descending);
        Assert.Null(sort.Toggle(CourseColumn.Title).Column);
        Assert.Equal(new CourseSortOrder(CourseColumn.Rating), sort.Toggle(CourseColumn.Rating));
    }

    [Fact]
    public void FiltersCombine()
    {
        var result = new CourseFilter { OnlyOpen = true, OnlyUnassigned = true, Query = "物理" }.Apply([
            Course("a", "大学物理", open: true),
            Course("b", "大学物理", open: true, group: "g"),
            Course("c", "大学物理"),
            Course("d", "高等数学", open: true),
        ]);
        Assert.Equal(["a"], result.Select(c => c.JxbId));
    }
}

/// 真正起一个 ng_service(需要仓库里的 conda Python),驱动 AppStore 走一遍主要流程。
public sealed class ServiceIntegrationTests : IDisposable
{
    readonly string dataDir = Path.Combine(Path.GetTempPath(), "ng-test-" + Guid.NewGuid().ToString("N"));
    readonly BlockingCollection<Action> uiQueue = new();
    readonly Thread uiThread;

    public ServiceIntegrationTests()
    {
        Directory.CreateDirectory(dataDir);
        uiThread = new Thread(() =>
        {
            SynchronizationContext.SetSynchronizationContext(new QueueContext(uiQueue));
            foreach (var action in uiQueue.GetConsumingEnumerable()) action();
        }) { IsBackground = true };
        uiThread.Start();
    }

    public void Dispose()
    {
        uiQueue.CompleteAdding();
        try { Directory.Delete(dataDir, recursive: true); } catch (IOException) { }
    }

    sealed class QueueContext(BlockingCollection<Action> queue) : SynchronizationContext
    {
        public override void Post(SendOrPostCallback d, object? state)
        {
            if (!queue.IsAddingCompleted) queue.Add(() => d(state));
        }
    }

    static string? RepoRoot()
    {
        for (var dir = new DirectoryInfo(AppContext.BaseDirectory); dir is not null; dir = dir.Parent)
            if (File.Exists(Path.Combine(dir.FullName, "ng_service.py"))) return dir.FullName;
        return null;
    }

    Task<T> OnUi<T>(Func<Task<T>> func)
    {
        var tcs = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
        uiQueue.Add(async () =>
        {
            try { tcs.SetResult(await func()); }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    static async Task WaitUntil(Func<bool> condition, int timeoutMs = 30_000)
    {
        var deadline = DateTime.UtcNow.AddMilliseconds(timeoutMs);
        while (!condition())
        {
            if (DateTime.UtcNow > deadline) throw new TimeoutException();
            await Task.Delay(50);
        }
    }

    [Fact]
    public async Task StoreDrivesRealService()
    {
        var root = RepoRoot();
        if (root is null) return; // 仓库外运行

        var term = Path.Combine(dataDir, "terms", "2026-3");
        Directory.CreateDirectory(term);
        File.WriteAllText(Path.Combine(dataDir, "user_settings.json"), """
            {"term":{"xkxnm":"2026","xkxqm":"3"},
             "terms":{"2026-3":{"priority_groups":{"物理":{"is_pe":false,"priority":["a","held"]}}}}}
            """);
        File.WriteAllText(Path.Combine(term, "catalog.json"), """
            {"courses":[{"kch":"PHY1","kcmc":"大学物理","kch_id":"K1","kklxdm":"01","classes":[
               {"jxb_id":"a","jxbmc":"物理-01","sksj":"星期一第2-3节{1-8周}"},
               {"jxb_id":"held","jxbmc":"物理-02","sksj":"星期二第1-2节{1-16周}"},
               {"jxb_id":"b","jxbmc":"物理-03","sksj":"星期一第1-2节{1-16周}"}]}],
             "choosed":[
               {"jxb_id":"held","kcmc":"大学物理","jxbmc":"物理-02","sksj":"星期二第1-2节{1-16周}"},
               {"jxb_id":"x","kcmc":"高等数学","jxbmc":"数学-01","sksj":"星期一第1-2节{1-16周}"}]}
            """);

        var env = System.Environment.GetEnvironmentVariables().Cast<System.Collections.DictionaryEntry>()
            .ToDictionary(e => (string)e.Key, e => (string?)e.Value ?? "", StringComparer.OrdinalIgnoreCase);
        env["SJTU_MONITOR_ROOT"] = root;
        BackendLaunch launch;
        try
        {
            launch = BackendLocator.Resolve(env);
        }
        catch (BackendException) when (!env.ContainsKey("SJTU_MONITOR_PYTHON"))
        {
            return; // 没有可用的 Python;显式设置了 SJTU_MONITOR_PYTHON 时必须真正跑起来
        }
        var launchEnv = new Dictionary<string, string>(launch.Environment) { ["SJTU_MONITOR_DATA_DIR"] = dataDir };
        launch = launch with { Environment = launchEnv };

        var notices = new ConcurrentQueue<Notice>();
        var store = new AppStore(action => uiQueue.Add(action), () => launch);
        store.NoticeRaised += notices.Enqueue;
        uiQueue.Add(store.Start);
        try
        {
            await WaitUntil(() => store.Snapshot is not null || store.Phase == BackendPhase.Failed);
            Assert.True(store.Phase == BackendPhase.Ready, store.Failure);
            Assert.Equal(BackendPhase.Ready, store.Phase);
            Assert.Equal("物理", store.SelectedGroup);
            Assert.Equal(1, store.Groups[0].ConflictCount);
            Assert.False(store.GroupsDirty);

            await OnUi(async () => { await store.AddCoursesAsync(["b"]); return true; });
            Assert.Equal(["a", "b", "held"], store.Groups[0].Priority);
            Assert.True(store.GroupsDirty);
            Assert.Contains(notices, n => n.Title == "时间冲突提示");
            await WaitUntil(() => store.Evaluations.TryGetValue("物理", out var e) && e.ConflictCount == 2);

            await OnUi(async () => { await store.SaveGroupsAsync(); return true; });
            Assert.False(store.GroupsDirty);
            Assert.Contains("\"b\"", File.ReadAllText(Path.Combine(dataDir, "user_settings.json")));

            await OnUi(async () => { await store.RunAsync("no-such-task"); return true; });
            Assert.Contains("未知任务", store.Status);

            await OnUi(async () => { await store.EnterDemoAsync(); return true; });
            Assert.True(store.DemoMode);
            Assert.Contains(store.Courses, c => c.JxbId.StartsWith("DEMO-"));
            await OnUi(async () => { await store.RunAsync("monitor"); return true; });
            Assert.Contains("演示模式", store.Status);
            await OnUi(async () => { await store.ExitDemoAsync(); return true; });
            Assert.False(store.DemoMode);
            Assert.Equal("物理", store.Groups[0].Name);
        }
        finally
        {
            store.Shutdown();
        }
    }
}
